# Mini Analytics Project

## Customer Purchase Behavior and Sales Performance

### Project Objective

The objective of this project is to analyze customer purchase behavior and sales performance using customer, order, product, order item, and transaction data. The analysis will identify the main factors driving revenue, highlight important business patterns, and provide recommendations based on the findings.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

pd.set_option("display.max_columns", None)

print("Libraries imported successfully.")

Libraries imported successfully.


In [4]:
from pathlib import Path

data_path = Path("../data/processed")

customers = pd.read_parquet(data_path / "clean_customers.parquet")
orders = pd.read_parquet(data_path / "clean_orders.parquet")
order_items = pd.read_parquet(data_path / "clean_order_items.parquet")
products = pd.read_parquet(data_path / "clean_products.parquet")
transactions = pd.read_parquet(data_path / "clean_transactions.parquet")

print("Customers:", customers.shape)
print("Orders:", orders.shape)
print("Order items:", order_items.shape)
print("Products:", products.shape)
print("Transactions:", transactions.shape)

FileNotFoundError: [Errno 2] No such file or directory: '..\\data\\processed\\clean_customers.parquet'

In [3]:
from pathlib import Path

print("Current folder:", Path.cwd())

Current folder: c:\Users\homam\OneDrive\Desktop\intern-bootcamp\05_analytics


In [5]:
from pathlib import Path

data_path = Path("../data/processed")

for file in data_path.iterdir():
    print(file.name)

customers_clean.parquet
features.csv
features_large.csv
logs
orders_clean.parquet
order_items_clean.parquet
products_clean.parquet
transactions_clean.parquet


In [6]:
from pathlib import Path

data_path = Path("../data/processed")

customers = pd.read_parquet(data_path / "customers_clean.parquet")
orders = pd.read_parquet(data_path / "orders_clean.parquet")
order_items = pd.read_parquet(data_path / "order_items_clean.parquet")
products = pd.read_parquet(data_path / "products_clean.parquet")
transactions = pd.read_parquet(data_path / "transactions_clean.parquet")

print("Customers:", customers.shape)
print("Orders:", orders.shape)
print("Order items:", order_items.shape)
print("Products:", products.shape)
print("Transactions:", transactions.shape)

Customers: (100, 5)
Orders: (200, 4)
Order items: (496, 6)
Products: (30, 4)
Transactions: (200, 6)


In [7]:
print("CUSTOMERS")
print(customers.columns.tolist())

print("\nORDERS")
print(orders.columns.tolist())

print("\nORDER ITEMS")
print(order_items.columns.tolist())

print("\nPRODUCTS")
print(products.columns.tolist())

print("\nTRANSACTIONS")
print(transactions.columns.tolist())

CUSTOMERS
['customer_id', 'first_name', 'email', 'city', 'signup_date']

ORDERS
['order_id', 'customer_id', 'order_date', 'status']

ORDER ITEMS
['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price', 'total_price']

PRODUCTS
['product_id', 'product_name', 'category', 'price']

TRANSACTIONS
['transaction_id', 'order_id', 'payment_method', 'payment_status', 'transaction_date', 'amount']


In [8]:
print("CUSTOMERS")
print(customers.dtypes)

print("\nORDERS")
print(orders.dtypes)

print("\nTRANSACTIONS")
print(transactions.dtypes)

CUSTOMERS
customer_id             int64
first_name                str
email                     str
city                      str
signup_date    datetime64[us]
dtype: object

ORDERS
order_id                int64
customer_id             int64
order_date     datetime64[us]
status                    str
dtype: object

TRANSACTIONS
transaction_id               int64
order_id                     int64
payment_method                 str
payment_status                 str
transaction_date    datetime64[us]
amount                     float64
dtype: object


In [9]:
analytics_data = (
    order_items
    .merge(orders, on="order_id", how="left")
    .merge(customers, on="customer_id", how="left")
    .merge(products, on="product_id", how="left")
    .merge(transactions, on="order_id", how="left")
)

print("Combined dataset shape:", analytics_data.shape)
analytics_data.head()

Combined dataset shape: (496, 21)


,order_item_id,order_id,product_id,quantity,unit_price,total_price,customer_id,order_date,status,first_name,email,city,signup_date,product_name,category,price,transaction_id,payment_method,payment_status,transaction_date,amount
0,1,1,27,5,216.77,1083.85,25,2026-01-04 17:14:47.555098,completed,Customer25,customer25@example.com,Sharjah,2025-09-06 17:14:47.543032,Product27,Books,216.77,1,card,paid,2026-01-04 17:14:47.555098,2467.21
1,2,1,29,4,345.84,1383.36,25,2026-01-04 17:14:47.555098,completed,Customer25,customer25@example.com,Sharjah,2025-09-06 17:14:47.543032,Product29,Books,345.84,1,card,paid,2026-01-04 17:14:47.555098,2467.21
2,3,2,5,2,260.38,520.76,27,2026-02-08 17:14:47.555110,completed,Customer27,customer27@example.com,Sharjah,2025-06-20 17:14:47.543036,Product5,Clothing,260.38,2,cash,failed,2026-02-08 17:14:47.555110,3263.99
3,4,2,29,3,345.84,1037.52,27,2026-02-08 17:14:47.555110,completed,Customer27,customer27@example.com,Sharjah,2025-06-20 17:14:47.543036,Product29,Books,345.84,2,cash,failed,2026-02-08 17:14:47.555110,3263.99
4,5,2,21,3,463.23,1389.69,27,2026-02-08 17:14:47.555110,completed,Customer27,customer27@example.com,Sharjah,2025-06-20 17:14:47.543036,Product21,Clothing,463.23,2,cash,failed,2026-02-08 17:14:47.555110,3263.99


In [10]:
print("Combined dataset shape:", analytics_data.shape)

Combined dataset shape: (496, 21)


In [11]:
analytics_data.isnull().sum()

order_item_id       0
order_id            0
product_id          0
quantity            0
unit_price          0
total_price         0
customer_id         0
order_date          0
status              0
first_name          0
email               0
city                0
signup_date         0
product_name        0
category            0
price               0
transaction_id      0
payment_method      0
payment_status      0
transaction_date    0
amount              0
dtype: int64

In [12]:
print("ORDER STATUS")
print(orders["status"].value_counts())

print("\nPAYMENT STATUS")
print(transactions["payment_status"].value_counts())

ORDER STATUS
status
completed    70
cancelled    66
pending      64
Name: count, dtype: int64

PAYMENT STATUS
payment_status
refunded    71
paid        67
failed      62
Name: count, dtype: int64


In [13]:
successful_orders = orders[orders["status"] == "completed"].copy()

successful_transactions = transactions[
    transactions["payment_status"] == "paid"
].copy()

print("Successful orders:", successful_orders.shape[0])
print("Successful transactions:", successful_transactions.shape[0])

Successful orders: 70
Successful transactions: 67


In [14]:
successful_order_ids = set(successful_orders["order_id"])
paid_order_ids = set(successful_transactions["order_id"])

fully_successful_order_ids = successful_order_ids.intersection(paid_order_ids)

print("Orders that are both completed and paid:", len(fully_successful_order_ids))

Orders that are both completed and paid: 23


In [15]:
successful_sales = analytics_data[
    analytics_data["order_id"].isin(fully_successful_order_ids)
].copy()

print("Successful sales rows:", successful_sales.shape[0])
print("Unique successful orders:", successful_sales["order_id"].nunique())

Successful sales rows: 57
Unique successful orders: 23


In [16]:
confirmed_revenue = successful_sales["total_price"].sum()

print(f"Confirmed revenue: {confirmed_revenue:,.2f}")

Confirmed revenue: 84,010.91


In [17]:
average_order_value = confirmed_revenue / len(fully_successful_order_ids)

print(f"Average order value: {average_order_value:,.2f}")

Average order value: 3,652.65


In [18]:
total_items_sold = successful_sales["quantity"].sum()

print("Total items sold:", total_items_sold)

Total items sold: 177


In [19]:
successful_order_rate = (
    len(fully_successful_order_ids) / orders["order_id"].nunique()
) * 100

print(f"Successful order rate: {successful_order_rate:.2f}%")

Successful order rate: 11.50%


In [20]:
successful_customer_orders = orders[
    orders["order_id"].isin(fully_successful_order_ids)
]

orders_per_customer = (
    successful_customer_orders
    .groupby("customer_id")["order_id"]
    .nunique()
)

returning_customers = (orders_per_customer > 1).sum()
successful_customers = orders_per_customer.shape[0]

returning_customer_rate = (
    returning_customers / successful_customers
) * 100

print("Successful customers:", successful_customers)
print("Returning customers:", returning_customers)
print(f"Returning customer rate: {returning_customer_rate:.2f}%")

Successful customers: 20
Returning customers: 3
Returning customer rate: 15.00%


In [21]:
payment_success_rate = (
    (transactions["payment_status"] == "paid").sum()
    / transactions["transaction_id"].nunique()
) * 100

print(f"Payment success rate: {payment_success_rate:.2f}%")

Payment success rate: 33.50%


In [22]:
kpi_summary = pd.DataFrame({
    "KPI": [
        "Confirmed Revenue",
        "Successful Orders",
        "Successful Order Rate",
        "Average Order Value",
        "Total Items Sold",
        "Successful Customers",
        "Returning Customer Rate",
        "Payment Success Rate"
    ],
    "Value": [
        f"{confirmed_revenue:,.2f}",
        len(fully_successful_order_ids),
        f"{successful_order_rate:.2f}%",
        f"{average_order_value:,.2f}",
        total_items_sold,
        successful_customers,
        f"{returning_customer_rate:.2f}%",
        f"{payment_success_rate:.2f}%"
    ]
})

kpi_summary

,KPI,Value
0,Confirmed Revenue,"84,010.91"
1,Successful Orders,23
2,Successful Order Rate,11.50%
3,Average Order Value,"3,652.65"
4,Total Items Sold,177
5,Successful Customers,20
6,Returning Customer Rate,15.00%
7,Payment Success Rate,33.50%


## Business Question 1: How does confirmed revenue change over time?

This analysis examines the daily revenue generated from orders that were both completed and successfully paid. The purpose is to identify whether revenue is stable, increasing, decreasing, or concentrated on specific days.

In [23]:
daily_revenue = (
    successful_sales
    .groupby("order_date", as_index=False)["total_price"]
    .sum()
    .sort_values("order_date")
)

daily_revenue.head()

,order_date,total_price
0,2025-11-21 17:14:47.555404,3804.16
1,2025-11-27 17:14:47.555388,650.31
2,2025-12-05 17:14:47.555406,5269.21
3,2025-12-11 17:14:47.555341,2754.87
4,2025-12-12 17:14:47.555209,4173.45


In [24]:
successful_sales["order_day"] = successful_sales["order_date"].dt.date

daily_revenue = (
    successful_sales
    .groupby("order_day", as_index=False)["total_price"]
    .sum()
    .sort_values("order_day")
)

daily_revenue.head()

,order_day,total_price
0,2025-11-21,3804.16
1,2025-11-27,650.31
2,2025-12-05,5269.21
3,2025-12-11,2754.87
4,2025-12-12,4173.45


In [25]:
fig = px.line(
    daily_revenue,
    x="order_day",
    y="total_price",
    markers=True,
    title="Daily Confirmed Revenue Trend"
)

fig.update_layout(
    xaxis_title="Order Date",
    yaxis_title="Confirmed Revenue"
)

fig.show()

In [26]:
highest_revenue_day = daily_revenue.loc[daily_revenue["total_price"].idxmax()]
lowest_revenue_day = daily_revenue.loc[daily_revenue["total_price"].idxmin()]

print("Highest revenue day:")
print(highest_revenue_day)

print("\nLowest revenue day:")
print(lowest_revenue_day)

Highest revenue day:
order_day      2026-05-10
total_price       9378.07
Name: 21, dtype: object

Lowest revenue day:
order_day      2026-04-15
total_price         203.1
Name: 15, dtype: object


### What does this chart mean?

The daily confirmed revenue is not stable across the analysis period. Revenue rises and falls significantly instead of following a consistent upward or downward trend.

The highest confirmed revenue was **9,378.07 on 10 May 2026**, while the lowest was **203.10 on 15 April 2026**. This shows that sales performance is concentrated on certain days.

From a business perspective, the company should investigate what caused the strongest revenue days, such as larger orders, popular products, promotions, or increased customer activity. Similar strategies could then be tested on weaker days to improve revenue consistency.

## Business Question 2: Which product categories generate the most confirmed revenue?

This analysis compares confirmed revenue across product categories. The purpose is to identify which categories contribute most to successful sales and which categories may need improvement.

In [27]:
category_revenue = (
    successful_sales
    .groupby("category", as_index=False)["total_price"]
    .sum()
    .sort_values("total_price", ascending=False)
)

category_revenue

,category,total_price
3,Home,25510.71
1,Clothing,21629.48
2,Electronics,19633.61
0,Books,17237.11


In [28]:
fig = px.bar(
    category_revenue,
    x="category",
    y="total_price",
    text_auto=".2s",
    title="Confirmed Revenue by Product Category"
)

fig.update_layout(
    xaxis_title="Product Category",
    yaxis_title="Confirmed Revenue"
)

fig.show()

### What does this chart mean?

The **Home** category generated the highest confirmed revenue at **25,510.71**, followed by **Clothing** at **21,629.48**.

**Books** generated the lowest confirmed revenue at **17,237.11**, although the difference between categories is not extremely large.

From a business perspective, the Home category is currently the strongest revenue contributor. The company should maintain good product availability in this category and investigate which Home products are driving the result. The weaker categories should also be reviewed to identify opportunities such as promotions, bundles, or better product positioning.

## Business Question 3: Which products generate the most confirmed revenue?

This analysis identifies the individual products contributing the most to confirmed revenue. The purpose is to determine which products are the strongest sales drivers within their categories.

In [29]:
product_revenue = (
    successful_sales
    .groupby("product_name", as_index=False)["total_price"]
    .sum()
    .sort_values("total_price", ascending=False)
    .head(10)
)

product_revenue

,product_name,total_price
12,Product23,8570.90
11,Product22,8448.90
2,Product11,7977.31
1,Product10,6961.05
9,Product20,6534.09
0,Product1,5503.12
14,Product25,5331.24
3,Product14,4882.14
16,Product27,4552.17
6,Product18,3361.26


In [30]:
fig = px.bar(
    product_revenue.sort_values("total_price"),
    x="total_price",
    y="product_name",
    orientation="h",
    text_auto=".2s",
    title="Top 10 Products by Confirmed Revenue"
)

fig.update_layout(
    xaxis_title="Confirmed Revenue",
    yaxis_title="Product"
)

fig.show()

### What does this chart mean?

**Product23** generated the highest confirmed revenue at **8,570.90**, followed closely by **Product22** at **8,448.90** and **Product11** at **7,977.31**.

The chart shows that a small group of products contributes a large share of confirmed revenue. These products are important sales drivers and should receive close attention.

From a business perspective, the company should maintain enough stock for the strongest products and investigate why customers prefer them. These products could also be used in promotions, recommendations, or bundles to increase sales of related products.

## Business Question 4: Which customer cities generate the most confirmed revenue?

This analysis compares confirmed revenue across customer cities. The purpose is to identify the locations containing the highest-value customers and the strongest sales opportunities.

In [31]:
city_revenue = (
    successful_sales
    .groupby("city", as_index=False)["total_price"]
    .sum()
    .sort_values("total_price", ascending=False)
)

city_revenue

,city,total_price
0,Abu Dhabi,26582.41
1,Ajman,26091.77
2,Dubai,18759.71
3,Sharjah,12577.02


In [32]:
fig = px.bar(
    city_revenue,
    x="city",
    y="total_price",
    text_auto=".2s",
    title="Confirmed Revenue by Customer City"
)

fig.update_layout(
    xaxis_title="Customer City",
    yaxis_title="Confirmed Revenue"
)

fig.show()

### What does this chart mean?

**Abu Dhabi** generated the highest confirmed revenue at **26,582.41**, followed closely by **Ajman** at **26,091.77**.

**Sharjah** generated the lowest confirmed revenue at **12,577.02**, which is much lower than Abu Dhabi and Ajman.

From a business perspective, Abu Dhabi and Ajman appear to contain the strongest customer markets. The company could focus more marketing and customer-retention efforts in these cities. Sharjah should also be investigated to understand whether lower revenue is caused by fewer customers, lower order values, or weaker product demand.

## Business Question 5: What is the distribution of payment outcomes?

This analysis compares paid, failed, and refunded transactions. The purpose is to understand whether payment problems may be reducing successful sales and confirmed revenue.

In [33]:
payment_distribution = (
    transactions["payment_status"]
    .value_counts()
    .rename_axis("payment_status")
    .reset_index(name="transaction_count")
)

payment_distribution

,payment_status,transaction_count
0,refunded,71
1,paid,67
2,failed,62


In [34]:
fig = px.bar(
    payment_distribution,
    x="payment_status",
    y="transaction_count",
    text="transaction_count",
    title="Transaction Count by Payment Status"
)

fig.update_layout(
    xaxis_title="Payment Status",
    yaxis_title="Number of Transactions"
)

fig.show()

### What does this chart mean?

Only **67 transactions were successfully paid**, while **71 were refunded** and **62 failed**.

This means that most transactions did not result in a successful payment. The payment success rate is only **33.50%**, which may be limiting confirmed revenue and the number of fully successful orders.

From a business perspective, the company should investigate the main causes of failed payments and refunds. Possible areas to review include payment-method reliability, checkout problems, customer cancellations, refund reasons, and transaction-processing issues.

## Business Question 6: How many successful customers returned to purchase again?

This analysis compares one-time customers with returning customers. The purpose is to understand whether the business is successfully retaining customers after their first completed and paid order.

In [35]:
customer_retention = pd.DataFrame({
    "customer_type": ["One-time customers", "Returning customers"],
    "customer_count": [
        successful_customers - returning_customers,
        returning_customers
    ]
})

customer_retention

,customer_type,customer_count
0,One-time customers,17
1,Returning customers,3


In [36]:
fig = px.bar(
    customer_retention,
    x="customer_type",
    y="customer_count",
    text="customer_count",
    title="Successful Customers by Customer Type"
)

fig.update_layout(
    xaxis_title="Customer Type",
    yaxis_title="Number of Customers"
)

fig.show()

### What does this chart mean?

Out of **20 successful customers**, **17 purchased only once** and only **3 returned to make another successful purchase**.

This gives a returning-customer rate of **15%**, which means the business currently depends mostly on one-time customers.

From a business perspective, the company should focus on customer-retention strategies such as loyalty rewards, follow-up offers, personalized recommendations, and promotions for previous customers. Increasing repeat purchases could improve long-term revenue without depending only on acquiring new customers.

## Key Findings

The analysis produced several important findings:

- Confirmed revenue reached **84,010.91** from **23 completed and paid orders**.
- The successful order rate was only **11.50%**.
- The payment success rate was **33.50%**, with many transactions failing or being refunded.
- The **Home** category generated the highest confirmed revenue.
- **Product23**, **Product22**, and **Product11** were the strongest individual products.
- **Abu Dhabi** and **Ajman** generated the most confirmed revenue.
- Daily revenue changed significantly, showing inconsistent sales performance.
- Only **15%** of successful customers returned to make another successful purchase.

## Business Recommendations

Based on the analysis, the business should:

- Investigate the causes of failed and refunded transactions to improve the payment success rate.
- Introduce loyalty programs, follow-up offers, or personalized recommendations to increase repeat purchases.
- Maintain strong product availability for the Home category and the highest-performing products.
- Focus marketing efforts on Abu Dhabi and Ajman, where confirmed revenue is strongest.
- Review the highest-revenue days to understand which products, customers, or activities caused the increase.

## Conclusion

This mini analytics project combined data preparation, KPI measurement, visualization, and business interpretation.

The results show that the business generated **84,010.91** in confirmed revenue, but only a small percentage of orders were both completed and successfully paid. Payment performance and customer retention are the main areas that need improvement.

The strongest revenue opportunities come from the Home category, the highest-performing products, and customers in Abu Dhabi and Ajman. Improving payment success, encouraging repeat purchases, and learning from the strongest sales days could help the business achieve more stable and sustainable revenue.